File: source/Photogrammetry/Reconstruction/TextureBaker.swift
import UIKit
import simd

/// Projects camera keyframes onto the UV atlas to produce a baked texture.
///
/// For each atlas texel the baker:
///   1. Identifies which face owns its cell (via `UVAtlasLayout`).
///   2. Computes barycentric coordinates within the fixed cell triangle
///      (v0=top-left, v1=top-right, v2=bottom-centre — matching `UVAtlasPacker`).
///   3. Interpolates the world-space position.
///   4. Finds the keyframe with the best viewing angle for that face.
///   5. Projects the world point through the keyframe camera and samples YCbCr.
///
/// Texels outside the cell triangle or with no valid keyframe get a mid-gray fill.
enum TextureBaker {
    /// Returns RGBA pixel data for a `layout.atlasSize × layout.atlasSize` image.
    static func bake(
        worldFaceVertices: [[SIMD3<Float>]],
        layout: UVAtlasLayout,
        keyframes: [CameraKeyframe]
    ) -> Data {
        let size = layout.atlasSize
        let cellSize = layout.cellSize
        let faceCount = worldFaceVertices.count

        var pixels = [UInt8](repeating: 128, count: size * size * 4)
        // Pre-fill alpha to 255 so the PNG is fully opaque.
        for i in stride(from: 3, to: pixels.count, by: 4) { pixels[i] = 255 }

        // Precompute per-keyframe data to avoid redundant work in the inner loop.
        let camInverses = keyframes.map { $0.transform.inverse }
        let camPositions = keyframes.map { kf -> SIMD3<Float> in
            SIMD3(kf.transform.columns.3.x,
                  kf.transform.columns.3.y,
                  kf.transform.columns.3.z)
        }

        for faceIndex in 0..<faceCount {
            let verts = worldFaceVertices[faceIndex]
            let origin = layout.cellOrigins[faceIndex]

            let e1 = verts[1] - verts[0]
            let e2 = verts[2] - verts[0]
            let faceNormal = normalize(cross(e1, e2))

            for ty in 0..<cellSize {
                for tx in 0..<cellSize {
                    // Barycentric coords for (tx,ty) within the fixed cell triangle.
                    // Derived from pA=(0,0), pB=(1,0), pC=(0.5,1) (normalised).
                    let tNorm = (Float(tx) + 0.5) / Float(cellSize)
                    let sNorm = (Float(ty) + 0.5) / Float(cellSize)
                    let λ0 = 1 - tNorm - 0.5 * sNorm
                    let λ1 = tNorm - 0.5 * sNorm
                    let λ2 = sNorm
                    guard λ0 >= 0, λ1 >= 0, λ2 >= 0 else { continue }

                    let worldPos = λ0 * verts[0] + λ1 * verts[1] + λ2 * verts[2]

                    let rgb = bestKeyframeSample(
                        worldPos: worldPos, faceNormal: faceNormal,
                        keyframes: keyframes, camInverses: camInverses,
                        camPositions: camPositions)
                    guard let rgb else { continue }

                    let atlasX = origin.x + tx
                    let atlasY = origin.y + ty
                    guard atlasX < size, atlasY < size else { continue }
                    let off = (atlasY * size + atlasX) * 4
                    pixels[off] = rgb.x; pixels[off+1] = rgb.y
                    pixels[off+2] = rgb.z; pixels[off+3] = 255
                }
            }
        }
        DiagnosticLog.info("TextureBaker — baked \(faceCount) faces")
        return Data(pixels)
    }

    /// Converts the raw RGBA data to a PNG for writing to disk.
    static func pngData(rgba: Data, size: Int) -> Data? {
        var mutableRGBA = rgba
        return mutableRGBA.withUnsafeMutableBytes { buf -> Data? in
            let colorSpace = CGColorSpaceCreateDeviceRGB()
            guard let ctx = CGContext(
                data: buf.baseAddress, width: size, height: size,
                bitsPerComponent: 8, bytesPerRow: size * 4,
                space: colorSpace,
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue),
                  let cgImg = ctx.makeImage() else { return nil }
            return UIImage(cgImage: cgImg).pngData()
        }
    }

    // MARK: - Private

    private static func bestKeyframeSample(
        worldPos: SIMD3<Float>,
        faceNormal: SIMD3<Float>,
        keyframes: [CameraKeyframe],
        camInverses: [simd_float4x4],
        camPositions: [SIMD3<Float>]
    ) -> SIMD3<UInt8>? {
        var bestDot: Float = 0
        var bestRGB: SIMD3<UInt8>? = nil

        for (i, keyframe) in keyframes.enumerated() {
            // Visibility: face normal must point toward the camera.
            let toCamera = normalize(camPositions[i] - worldPos)
            let facing = dot(faceNormal, toCamera)
            guard facing > bestDot else { continue }

            // Project world position into camera space (ARKit: camera looks down -Z).
            let camSpace = camInverses[i] * SIMD4<Float>(worldPos, 1)
            guard camSpace.z < 0 else { continue }

            // camera.transform is sensor-aligned (same convention as
            // ARCamera.intrinsics), so camSpace.x/y map directly to
            // sensor U/V. No orientation rotation is needed here.
            let k = keyframe.intrinsics
            let projX = k[2][0] + k[0][0] * camSpace.x / (-camSpace.z)
            let projY = k[2][1] + k[1][1] * camSpace.y / (-camSpace.z)
            let imgW = Float(keyframe.imageResolution.width)
            let imgH = Float(keyframe.imageResolution.height)
            guard projX >= 0, projX < imgW, projY >= 0, projY < imgH else { continue }

            bestDot = facing
            bestRGB = YCbCrSampler.sampleRGB(
                keyframe.image, u: projX / imgW, v: projY / imgH)
        }
        return bestRGB
    }
}

File: server/internal/processing/texture_baker.go
package processing

import (
	"image"
	"image/color"
)

// Keyframe is a decoded capture frame plus its camera parameters.
// Transform is the camera-to-world matrix (column-major). Intrinsics is the
// 3x3 camera matrix, column-major (element = col*3+row).
type Keyframe struct {
	Image      image.Image
	Transform  Mat4
	Intrinsics [9]float32
	Width      int
	Height     int
}

// fx, fy, cx, cy extract focal lengths and principal point from the
// column-major 3x3 intrinsics (fx@0, fy@4, cx@6, cy@7).
func (k Keyframe) fx() float32 { return k.Intrinsics[0] }
func (k Keyframe) fy() float32 { return k.Intrinsics[4] }
func (k Keyframe) cx() float32 { return k.Intrinsics[6] }
func (k Keyframe) cy() float32 { return k.Intrinsics[7] }

// camPrep holds per-keyframe precomputed projection state.
type camPrep struct {
	inverse  Mat4       // world-to-camera
	position [3]float32 // camera world position (transform col 3)
}

// BakeTexture renders a baked color atlas by projecting each atlas texel onto
// the best front-facing keyframe and sampling its color. Ports the Swift
// TextureBaker. Returns an *image.RGBA of atlasSize×atlasSize.
func BakeTexture(faces [][3][3]float32, layout UVAtlasLayout, keyframes []Keyframe) *image.RGBA {
	atlas := newFilledAtlas(layout.AtlasSize)
	preps := prepCameras(keyframes)
	for f := range faces {
		bakeFace(atlas, faces[f], layout.CellOrigins[f], layout.CellSize, keyframes, preps)
	}
	return atlas
}

// newFilledAtlas allocates an RGBA atlas pre-filled with RGB=128, A=255.
func newFilledAtlas(size int) *image.RGBA {
	atlas := image.NewRGBA(image.Rect(0, 0, size, size))
	for i := 0; i < len(atlas.Pix); i += 4 {
		atlas.Pix[i], atlas.Pix[i+1], atlas.Pix[i+2], atlas.Pix[i+3] = 128, 128, 128, 255
	}
	return atlas
}

// prepCameras precomputes the world-to-camera inverse and camera world
// position for each keyframe.
func prepCameras(keyframes []Keyframe) []camPrep {
	preps := make([]camPrep, len(keyframes))
	for i, kf := range keyframes {
		preps[i] = camPrep{
			inverse:  Mat4Inverse(kf.Transform),
			position: [3]float32{kf.Transform[12], kf.Transform[13], kf.Transform[14]},
		}
	}
	return preps
}

// bakeFace rasterizes one face's cell, sampling each texel from the best
// front-facing keyframe.
func bakeFace(atlas *image.RGBA, verts [3][3]float32, origin [2]int, cellSize int, keyframes []Keyframe, preps []camPrep) {
	e1 := vec3Sub(verts[1], verts[0])
	e2 := vec3Sub(verts[2], verts[0])
	normal := vec3Normalize(vec3Cross(e1, e2))
	for ty := 0; ty < cellSize; ty++ {
		for tx := 0; tx < cellSize; tx++ {
			world, ok := texelWorldPos(verts, tx, ty, cellSize)
			if !ok {
				continue
			}
			if rgb, ok := bestKeyframeColor(world, normal, keyframes, preps); ok {
				setAtlasTexel(atlas, origin[0]+tx, origin[1]+ty, rgb)
			}
		}
	}
}

// texelWorldPos converts a cell-local texel to a world position via the
// barycentric scheme from the Swift baker; ok=false when outside the triangle.
func texelWorldPos(verts [3][3]float32, tx, ty, cellSize int) ([3]float32, bool) {
	tNorm := (float32(tx) + 0.5) / float32(cellSize)
	sNorm := (float32(ty) + 0.5) / float32(cellSize)
	l0 := 1 - tNorm - 0.5*sNorm
	l1 := tNorm - 0.5*sNorm
	l2 := sNorm
	if l0 < 0 || l1 < 0 || l2 < 0 {
		return [3]float32{}, false
	}
	var w [3]float32
	for i := 0; i < 3; i++ {
		w[i] = l0*verts[0][i] + l1*verts[1][i] + l2*verts[2][i]
	}
	return w, true
}

// bestKeyframeColor selects the most front-facing keyframe that sees worldPos
// in front of the camera and within image bounds, and returns its sampled RGB.
func bestKeyframeColor(world, normal [3]float32, keyframes []Keyframe, preps []camPrep) (color.RGBA, bool) {
	var bestDot float32
	var best color.RGBA
	found := false
	for i := range keyframes {
		facing := vec3Dot(normal, vec3Normalize(vec3Sub(preps[i].position, world)))
		if facing <= bestDot {
			continue
		}
		rgb, ok := projectAndSample(world, keyframes[i], preps[i])
		if !ok {
			continue
		}
		bestDot = facing
		best = rgb
		found = true
	}
	return best, found
}

// projectAndSample projects worldPos into keyframe kf and returns the sampled
// color, or ok=false if behind the camera or off-image.
func projectAndSample(world [3]float32, kf Keyframe, prep camPrep) (color.RGBA, bool) {
	cam := Mat4MulVec(prep.inverse, [4]float32{world[0], world[1], world[2], 1})
	if cam[2] >= 0 { // ARKit camera looks down -Z; positive Z is behind
		return color.RGBA{}, false
	}
	projX := kf.cx() + kf.fx()*cam[0]/(-cam[2])
	projY := kf.cy() + kf.fy()*cam[1]/(-cam[2])
	if projX < 0 || projX >= float32(kf.Width) || projY < 0 || projY >= float32(kf.Height) {
		return color.RGBA{}, false
	}
	return sampleNearest(kf.Image, int(projX), int(projY)), true
}

// sampleNearest reads the nearest-neighbor pixel as an opaque RGBA color.
func sampleNearest(img image.Image, x, y int) color.RGBA {
	r, g, b, _ := img.At(img.Bounds().Min.X+x, img.Bounds().Min.Y+y).RGBA()
	return color.RGBA{uint8(r >> 8), uint8(g >> 8), uint8(b >> 8), 255}
}

// setAtlasTexel writes an opaque color into the atlas at (x, y).
func setAtlasTexel(atlas *image.RGBA, x, y int, c color.RGBA) {
	atlas.SetRGBA(x, y, c)
}
